Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/FilterAdapter.cpp

```diff
@@ -368,15 +360,144 @@ namespace AZ::DocumentPropertyEditor
         }
     }
 
-    void RowFilterAdapter::UpdateMatchState(MatchInfoNode* rowState)
+    void RowFilterAdapter::GenerateFullTree()
     {
-        auto includeInFilter = [](MatchInfoNode* node) -> bool
+        AZ_Assert(!m_root, "GenerateFullTree must only be called on an empty tree!");
+        m_root = MakeNewNode(nullptr, m_updateFrame);
+        const auto& sourceContents = m_sourceAdapter->GetContents();
+
+        // we can assume all direct children of an adapter must be rows; populate each of them
+        for (size_t topIndex = 0; topIndex < sourceContents.ArraySize(); ++topIndex)
+        {
+            PopulateNodesAtPath(Dom::Path({ Dom::PathEntry(topIndex) }), false);
+        }
+    }
+
+    void RowFilterAdapter::CullUnmatchedChildRows(Dom::Value& rowValue, const MatchInfoNode* rowMatchNode)
+    {
+        auto& matchChildren = rowMatchNode->m_childMatchState;
+        const bool sizesMatch = (rowValue.ArraySize() == matchChildren.size());
+        AZ_Assert(sizesMatch, "DOM value and cached match node should have the same number of children!");
+
+        if (sizesMatch)
+        {
+            auto valueIter = rowValue.MutableArrayBegin();
+            for (auto matchIter = matchChildren.begin(); matchIter != matchChildren.end(); ++matchIter)
+            {
+                // non-row children will have a null MatchInfo -- we only need to worry about culling row values
+                auto currMatch = *matchIter;
+                if (currMatch != nullptr)
+                {
+                    if (currMatch->IncludeInFilter())
+                    {
+                        CullUnmatchedChildRows(*valueIter, *matchIter);
+                    }
+                    else
+                    {
+                        valueIter = rowValue.ArrayErase(valueIter);
+                        continue; // we've already moved valueIter forward, skip the ++valueIter below
+                    }
+                }
+                ++valueIter;
+            }
+        }
+    }
+
+    void RowFilterAdapter::GeneratePatch(PatchType patchType)
+    {
+        Dom::Patch patch;
+
+        AZStd::function<void(MatchInfoNode*, Dom::Path, const Dom::Value&)> generatePatchOperations =
+            [&](MatchInfoNode* matchNode, Dom::Path mappedPath, const Dom::Value& sourceValue)
```

Comment:
No need to store this in an AZStd::function prevents the `operator()` call on the lambda from being inlined.
`auto` works in this case because the lambda `operator()` can be invoked directly.
```suggestion
        auto generatePatchOperations =
            [&](MatchInfoNode* matchNode, Dom::Path mappedPath, const Dom::Value& sourceValue)
```

Reply:
I tried that, but since it can call itself recursively on line 439, auto isn't working for me. It gives me the error:
`error C3536: 'generatePatchOperations': cannot be used before it is initialized`
Is there another way around that rather than storing it in a named function?

Reply:
Well you could make a struct with an `operator()` which would allow to avoid the auto?

Reply:
This is quite an interesting topic, so I did a deep dive on recursive lambdas and found this:
https://artificial-mind.net/blog/2020/09/12/recursive-lambdas

I followed this which worked as long as I gave the lambda a trailing return type, which I did. See if you like it now...

Reply:
That works as well